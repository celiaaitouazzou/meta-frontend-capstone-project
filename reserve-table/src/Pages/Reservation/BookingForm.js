import React from 'react';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { Button, Row, Col, Container } from 'react-bootstrap'; // Still using Container, Row, Col for layout
import greekSalad from './GreekSalad.jpg';

function BookingForm(props) {
  return (
    <Container>
      <Row style={{ backgroundColor: '#F4CE14', padding: '5%' }}>
        <Col style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-evenly', alignItems: 'baseline' }}>
          <Formik
            initialValues={props.formData || {
              firstName: '',
              lastName: '',
              date: '',
              time: '',
              guests: '1',
              occasion: 'Occasion',
            }}
            validate={values => {
              const errors = {};
              if (!values.firstName) {
                errors.firstName = 'Required';
              }
              if (!values.lastName) {
                errors.lastName = 'Required';
              }
              if (!values.date) {
                errors.date = 'Required';
              }
              if (!values.time) {
                errors.time = 'Required';
              }
              if (!values.guests) {
                errors.guests = 'Required';
              }
              return errors;
            }}
            onSubmit={(values, { setSubmitting }) => {
              setTimeout(() => {
                alert(JSON.stringify(values, null, 2));
                setSubmitting(false);
              }, 400);
            }}
          >
            {({ isSubmitting, values, touched, errors, handleChange, handleBlur }) => (
              <Form>
                <div className="mb-3">
                  <label htmlFor="firstName" className="form-label">
                    First Name
                  </label>
                  <input
                    type="text"
                    className={`form-control ${touched.firstName && errors.firstName ? 'is-invalid' : ''}`}
                    id="firstName"
                    name="firstName"
                    placeholder="First Name"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.firstName}
                  />
                  {touched.firstName && errors.firstName && (
                    <div className="invalid-feedback">{errors.firstName}</div>
                  )}
                </div>

                <div className="mb-3">
                  <label htmlFor="lastName" className="form-label">
                    Last Name
                  </label>
                  <input
                    type="text"
                    className={`form-control ${touched.lastName && errors.lastName ? 'is-invalid' : ''}`}
                    id="lastName"
                    name="lastName"
                    placeholder="Last Name"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.lastName}
                  />
                  {touched.lastName && errors.lastName && (
                    <div className="invalid-feedback">{errors.lastName}</div>
                  )}
                </div>

                <div className="mb-3">
                  <label htmlFor="date" className="form-label">
                    Date
                  </label>
                  <input
                    type="date"
                    className={`form-control ${touched.date && errors.date ? 'is-invalid' : ''}`}
                    id="date"
                    name="date"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.date}
                  />
                  {touched.date && errors.date && (
                    <div className="invalid-feedback">{errors.date}</div>
                  )}
                </div>

                <div className="mb-3">
                  <label htmlFor="time" className="form-label">
                    Time
                  </label>
                  <input
                    type="time"
                    className={`form-control ${touched.time && errors.time ? 'is-invalid' : ''}`}
                    id="time"
                    name="time"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.time}
                  />
                  <div className="form-text">We are open from 8:00 AM to 10 PM.</div>
                  {touched.time && errors.time && (
                    <div className="invalid-feedback">{errors.time}</div>
                  )}
                </div>

                <div className="mb-3">
                  <label htmlFor="guests" className="form-label">
                    Selected Number of Guests: {values.guests}
                  </label>
                  <input
                    type="range"
                    className={`form-range ${touched.guests && errors.guests ? 'is-invalid' : ''}`}
                    id="guests"
                    name="guests"
                    min="1"
                    max="12"
                    step="1"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.guests}
                  />
                  <div className="form-text">We can accommodate up to 12 people parties.</div>
                  {touched.guests && errors.guests && (
                    <div className="invalid-feedback">{errors.guests}</div>
                  )}
                </div>

                <div className="mb-3">
                  <label htmlFor="occasion" className="form-label">
                    Select an Occasion
                  </label>
                  <select
                    className="form-select"
                    id="occasion"
                    name="occasion"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.occasion}
                  >
                    <option value="Occasion">Select an Occasion</option>
                    <option value="Birthday">Birthday</option>
                    <option value="Engagement">Engagement</option>
                    <option value="Anniversary">Anniversary</option>
                  </select>
                </div>

                <Button variant="dark" type="submit" disabled={isSubmitting}>
                  Submit
                </Button>
              </Form>
            )}
          </Formik>
        </Col>
        <Col>
          <img src={greekSalad} width={800} height={600} alt="greek-salad" />
        </Col>
      </Row>
    </Container>
  );
}

export default BookingForm;