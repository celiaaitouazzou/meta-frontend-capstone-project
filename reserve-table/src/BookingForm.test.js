import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import BookingForm from './Pages/Reservation/BookingForm';

describe('BookingForm - Required Fields and Submission Prevention', () => {
  const mockFormData = {
    firstName: '',
    lastName: '',
    date: '',
    time: '',
    guests: 2,
    occasion: '',
  };

  const renderForm = (formData = mockFormData) => {
    render(<BookingForm formData={formData} />);
  };

  test('displays "Required" error for empty First Name and disables submit', () => {
    renderForm();
    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(screen.getByText('Required')).toBeVisible();
    expect(screen.getByLabelText('First Name').closest('.form-group')).toContainElement(screen.getByText('Required'));
    expect(screen.getByRole('button', { name: /Submit/i })).toBeDisabled();
  });

  test('displays "Required" error for empty Last Name and disables submit', () => {
    renderForm();
    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(screen.getByText('Required')).toBeVisible();
    expect(screen.getByLabelText('Last Name').closest('.form-group')).toContainElement(screen.getByText('Required'));
    expect(screen.getByRole('button', { name: /Submit/i })).toBeDisabled();
  });

  test('displays "Required" error for empty Date and disables submit', () => {
    renderForm();
    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(screen.getByText('Required')).toBeVisible();
    expect(screen.getByLabelText('Date').closest('.form-group')).toContainElement(screen.getByText('Required'));
    expect(screen.getByRole('button', { name: /Submit/i })).toBeDisabled();
  });

  test('displays "Required" error for empty Time and disables submit', () => {
    renderForm();
    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(screen.getByText('Required')).toBeVisible();
    expect(screen.getByLabelText('Time').closest('.form-group')).toContainElement(screen.getByText('Required'));
    expect(screen.getByRole('button', { name: /Submit/i })).toBeDisabled();
  });

  test('enables submit button and hides "Required" errors when all required fields are filled', () => {
    renderForm({
      firstName: 'John',
      lastName: 'Doe',
      date: '2025-05-08',
      time: '12:00',
      guests: 2,
      occasion: '',
    });

    fireEvent.focus(screen.getByLabelText('First Name'));
    fireEvent.blur(screen.getByLabelText('First Name'));
    fireEvent.focus(screen.getByLabelText('Last Name'));
    fireEvent.blur(screen.getByLabelText('Last Name'));
    fireEvent.focus(screen.getByLabelText('Date'));
    fireEvent.blur(screen.getByLabelText('Date'));
    fireEvent.focus(screen.getByLabelText('Time'));
    fireEvent.blur(screen.getByLabelText('Time'));

    expect(screen.queryByText('Required', { container: screen.getByLabelText('First Name').closest('.form-group') })).toBeNull();
    expect(screen.queryByText('Required', { container: screen.getByLabelText('Last Name').closest('.form-group') })).toBeNull();
    expect(screen.queryByText('Required', { container: screen.getByLabelText('Date').closest('.form-group') })).toBeNull();
    expect(screen.queryByText('Required', { container: screen.getByLabelText('Time').closest('.form-group') })).toBeNull();
    expect(screen.getByRole('button', { name: /Submit/i })).toBeEnabled();
  });

  test('does not call handleSubmit if required fields are empty', () => {
    const alertMock = jest.spyOn(window, 'alert').mockImplementation(() => {});
    renderForm();
    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(alertMock).not.toHaveBeenCalled();
    alertMock.mockRestore();
  });

  test('calls handleSubmit if all required fields are filled', () => {
    const alertMock = jest.spyOn(window, 'alert').mockImplementation(() => {});
    const validFormData = {
      firstName: 'John',
      lastName: 'Doe',
      date: '2025-05-08',
      time: '12:00',
      guests: 2,
      occasion: '',
    };
    renderForm(validFormData);

    fireEvent.focus(screen.getByLabelText('First Name'));
    fireEvent.blur(screen.getByLabelText('First Name'));
    fireEvent.focus(screen.getByLabelText('Last Name'));
    fireEvent.blur(screen.getByLabelText('Last Name'));
    fireEvent.focus(screen.getByLabelText('Date'));
    fireEvent.blur(screen.getByLabelText('Date'));
    fireEvent.focus(screen.getByLabelText('Time'));
    fireEvent.blur(screen.getByLabelText('Time'));

    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(alertMock).toHaveBeenCalledWith(JSON.stringify(validFormData, null, 2));
    alertMock.mockRestore();
  });
});