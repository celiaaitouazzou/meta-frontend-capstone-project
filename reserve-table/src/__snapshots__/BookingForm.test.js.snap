// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingForm renders without crashing and matches basic structure 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
      style="background-color: rgb(244, 206, 20); padding: 5%;"
    >
      <div
        class="col"
        style="display: flex; flex-direction: row; justify-content: space-evenly; align-items: baseline;"
      >
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="formBasicFN"
            >
              First Name
            </label>
            <input
              class="form-control"
              id="formBasicFN"
              name="firstName"
              placeholder="First Name"
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="formBasicLN"
            >
              Last Name
            </label>
            <input
              class="form-control"
              id="formBasicLN"
              name="lastName"
              placeholder="Last Name"
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="dateInput"
            >
              Date
            </label>
            <input
              class="form-control"
              id="dateInput"
              name="date"
              type="date"
              value=""
            />
            <div
              class="invalid-feedback"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="timeInput"
            >
              Time
            </label>
            <input
              class="form-control"
              id="timeInput"
              name="time"
              type="time"
              value=""
            />
            <small
              class="form-text text-muted"
            >
              We are open from 8:00 AM to 10 PM.
            </small>
            <div
              class="invalid-feedback"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              aria-live="polite"
              class="form-label"
              for="noGuest"
            >
              Selected Number of Guests: 
              1
            </label>
            <input
              class="form-control"
              id="noGuest"
              max="12"
              min="1"
              name="guests"
              step="1"
              type="range"
              value="1"
            />
            <small
              class="form-text text-muted"
            >
              We can accommodate up to 12 people parties.
            </small>
            <div
              class="invalid-feedback"
            />
          </div>
          <div
            class="mb-3"
          >
            <select
              class="form-select"
              name="occasion"
            >
              <option
                value="Occasion"
              >
                Select an Occasion
              </option>
              <option
                value="Birthday"
              >
                Birthday
              </option>
              <option
                value="Engagement"
              >
                Engagement
              </option>
              <option
                value="Anniversary"
              >
                Anniversary
              </option>
            </select>
          </div>
          <button
            class="btn btn-dark"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
      <div
        class="col"
      >
        <img
          alt="greek-salad"
          height="600"
          src="GreekSalad.jpg"
          width="800"
        />
      </div>
    </div>
  </div>
</div>
`;
